### PR TITLE
new: GET /entities/:uuid/audits

### LINKED ISSUE #796: Entity audit log
Related to dataset audit log issue: https://github.com/getodk/central-backend/issues/795 

Add `GET /v1/projects/:projectId/datasets/:name/entities/:uuid/audits` to show events about creating and updating a specific entity.

Questions 
- what will frontend need to show?
- what info do we need to store?
- should come after/along with work to create and update entities

### PR DESCRIPTION
**Pull request by sadiqkhoja**
Closes #796

#### Before submitting this PR, please make sure you have:

- [ ] run `make test-full` and confirmed all checks still pass OR confirm CircleCI build passes
- [ ] verified that any code from external sources are properly credited in comments or that everything is internally sourced

### CODE REVIEW

**Comment by sadiqkhoja** [comment]
Sample response:

```javascript
[
  {
    actorId: 6,
    action: "entity.update.version",
    acteeId: "a889f7cf-202c-4213-88a7-ca20dd61a1c7",
    details: {
      uuid: "12345678-1234-4123-8234-123456789abc",
      dataset: "people",
      entityId: 1,
      entityDefId: 3,
    },
    loggedAt: "2023-04-25T18:43:03.431Z",
    notes: null,
    actor: {
      id: 6,
      type: "user",
      displayName: "Bob",
      createdAt: "2023-04-25T18:43:02.560Z",
      updatedAt: null,
      deletedAt: null,
    },
  },
  {
    actorId: 5,
    action: "entity.create",
    acteeId: "a889f7cf-202c-4213-88a7-ca20dd61a1c7",
    details: {
      submission: {
        instanceId: "one",
        submitterId: 5,
        deviceId: null,
        createdAt: "2023-04-25T18:43:02.941Z",
        updatedAt: "2023-04-25T18:43:02.948Z",
        reviewState: "approved",
        userAgent: "node-superagent/3.8.3",
        submitter: {
          id: 5,
          type: "user",
          displayName: "Alice",
          createdAt: "2023-04-25T18:43:02.550Z",
          updatedAt: null,
          deletedAt: null,
        },
        currentVersion: {
          submitterId: 5,
          createdAt: "2023-04-25T18:43:02.941Z",
          instanceName: "one",
          instanceId: "one",
          current: true,
          deviceId: null,
          userAgent: "node-superagent/3.8.3",
          submitter: {
            id: 5,
            type: "user",
            displayName: "Alice",
            createdAt: "2023-04-25T18:43:02.550Z",
            updatedAt: null,
            deletedAt: null,
          },
        },
      },
      approval: {
        actorId: 5,
        action: "submission.update",
        acteeId: "3975f854-e47f-485e-b3a9-ef20cb4d9ed0",
        details: {
          reviewState: "approved",
          submissionId: 1,
          submissionDefId: 1,
        },
        loggedAt: "2023-04-25T18:43:02.949Z",
        notes: null,
        actor: {
          id: 5,
          type: "user",
          displayName: "Alice",
          createdAt: "2023-04-25T18:43:02.550Z",
          updatedAt: null,
          deletedAt: null,
        },
      },
      entity: {
        uuid: "12345678-1234-4123-8234-123456789abc",
        label: "Alice (88)",
        dataset: "people",
      },
      entityId: 1,
      submissionId: 1,
      submissionDefId: 1,
    },
    loggedAt: "2023-04-25T18:43:02.975Z",
    notes: null,
    actor: {
      id: 5,
      type: "user",
      displayName: "Alice",
      createdAt: "2023-04-25T18:43:02.550Z",
      updatedAt: null,
      deletedAt: null,
    },
  },
]
```

File: lib/model/query/audits.js
Comment (matthew-white) [inline lib/model/query/audits.js:124]: Unless I'm misunderstanding, I think we need to handle the possibility of multiple approvals of the same submission. I've been wondering about that case, and I'm not sure how to select the correct approval when there are more than one. Maybe we could get away with selecting the submission's first approval? I could imagine edge cases where that doesn't work, though they seems pretty unlikely. For example:

- A submission is created for a form that can create entities, but the submission doesn't specify an entity. (That's possible, right?)
- The submission is approved. No entity is created.
- Someone later realizes that there was a mistake and edits the submission. This time, the submission version does specify an entity.
- The submission is reapproved, creating the entity.

If we want to be more robust about cases like that, maybe we select the first approval whose `processed` timestamp is greater than or equal to the `loggedAt` timestamp of the `entity.create` event ? Or maybe we add a column to the entity or entity source that identifies the specific approval that created the entity?

I think the most important thing though is to handle multiple approvals in some way. I think multiple approvals are a real possibility and that we'll want to avoid returning duplicate audit log entries in that case.

File: lib/model/query/audits.js
Comment (matthew-white) [inline lib/model/query/audits.js:120]: What do you think about joining in `submissions` just for `entity.create`, not `entity.update.version`? At least right now, submissions can't update entities, so I think the only action that's really connected to submissions is `entity.create`.

**Review by matthew-white — COMMENTED** [review]
Just a couple of preliminary thoughts!

**Comment by matthew-white** [comment]
I'm realizing that Frontend will also need to learn the xmlFormId of the submission so it can link to the submission.

File: lib/model/query/entities.js
Comment (matthew-white) [inline lib/model/query/entities.js:100]: Do we need to backfill `entityId`? Or are we OK with the entity activity feed being empty for entities created before .3?

File: lib/model/query/entities.js
Comment (matthew-white) [inline lib/model/query/entities.js:101]: I think we're planning not to backfill this. If that's right, could you leave a comment above this property that notes that it wasn't saved for entities created before v2023.3? I think it'll be nice to have some record of how the details schema has changed over time.

**Review by matthew-white — COMMENTED** [review]
I'll let @ktuite do the main review, but I thought I'd leave a few comments. (I'm also happy to do the main review if @ktuite prefers.)

@sadiqkhoja and I have discussed the schema of the `details` object of the `entity.create` event. I think that's turned out nicely, and it provides all the information that Frontend needs for the entity activity feed and for "Creating Submission" in the Basic Information section of the entity detail page.

This PR requires storing an additional piece of data in the audit log, `eventId`, which points to the specific approval event that resulted in the entity's creation. I think that's a simple and robust approach to the problem of identifying the correct approval. I don't think we need to backfill `eventId`, since entities are still in preview. For entities created before .3, the entity activity feed won't show the submission approval event, but I think that's fine.

This is only sort of related and would happen later, but I'm coming around to @ktuite's idea of having one set of public audit log details that are returned over the API and another set that is only for internal use:

- Frontend only uses some details, and distinguishing between public details and internal ones might be a way to document which details are more stable (for example, those used by Frontend) and which ones are just to facilitate Backend's implementation.
- Distinguishing between public and internal details would be a way to prevent returning entity and submission numeric ids over the API. We generally don't expose those ids over the API.

File: lib/model/query/audits.js
Comment (sadiqkhoja) [reply lib/model/query/audits.js:130]: this is new to me as well. I learnt it from here: https://stackoverflow.com/a/56815807/3633589

Comment (matthew-white) [reply lib/model/query/audits.js:130]: Fascinating! Even posted by Gajus of Slonik fame.

File: lib/model/query/entities.js
Comment (matthew-white) [reply lib/model/query/entities.js:100]: We've decided not to backfill `entityId`. It feels best to avoid changing the audit log when possible so that it can serve as a historical record.

Similar to my comment below, would you mind leaving a comment in the code that mentions that this property was added in v2023.3?

File: test/integration/api/entities.js
Comment (sadiqkhoja) [inline test/integration/api/entities.js:329]: add tests to verify pagination query options

File: test/data/xml.js
Comment (sadiqkhoja) [inline test/data/xml.js:582]: add instanceName to other submissions of `simpleEntity`

File: test/integration/api/entities.js
Comment (sadiqkhoja) [inline test/integration/api/entities.js:415]: explain background of this test

File: test/integration/api/entities.js
Comment (sadiqkhoja) [inline test/integration/api/entities.js:469]: approve the submission one more time here

File: test/integration/api/entities.js
Comment (sadiqkhoja) [inline test/integration/api/entities.js:330]: add negative tests in here as well

File: test/integration/api/entities.js
Comment (sadiqkhoja) [inline test/integration/api/entities.js:439]: add a test with POST API

**Review by ktuite — APPROVED** [review]
I've had a chance to play around with this a bit on my own, and Sadiq has talked me through it.

**Comment by matthew-white** [comment]
I just had the thought, shouldn't we only return actor objects if extended metadata is requested?

No need to block merging (we could address that question in a follow-up PR), but I thought I'd put the question out there.

**Comment by ktuite** [comment]
> I just had the thought, shouldn't we only return actor objects if extended metadata is requested?
> 
> No need to block merging (we could address that question in a follow-up PR), but I thought I'd put the question out there.

Sadiq could comment more, but I think because of all the work the queries are doing to fetch and fill in the details, it's harder to distinguish between what extended vs. not should really mean. It might be tricky or not particularly useful to just collapse the actors within the different levels of nesting. Maybe extending could mean un-collapsing something in addition to the actors, or maybe not.

There are actors within
- the main audit event
- if the event is an entity create event
   - the approval event if the entity was created from a submission approval
   - the submission creation event if the entity was from a submission
   - the submitter of the submission itself that made the entity
- maybe more when the event is an entity update event with a submission as its source (not an api call)

Anyway... definitely worth thinking about but it shouldn't block this PR.

**Comment by sadiqkhoja** [comment]
My thinking regarding `extended` is that this endpoint is created to fulfil the need of frontend specifically. I don't foresee situation where we would need the response without Actor object extended

**Comment by ktuite** [comment]
Was staring at this again today and thinking... do we need `submissionCreate` in the details there? Is the Creator information on the submission itself enough?

**Comment by matthew-white** [comment]
I think it's reasonable to leave things the way they are. :+1: The endpoint will sort of return extended metadata by default. It doesn't feel like the typical Backend pattern: even the submission audit log can return extended metadata or not. But it also seems fine to have this exception, and it doesn't feel worth spending a lot of time to change how it's working.

> Was staring at this again today and thinking... do we need `submissionCreate` in the details there? Is the Creator information on the submission itself enough?

`submissionCreate` is needed in case the submission was deleted. Frontend can use `submissionCreate` for most things, but it needs the `submission` property for the instance name.

**Pull request merged**
